src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,168 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.NoSuchElementException;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    static class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> prev;
+
+        private Node(T element) {
+            this.element = element;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        }
+        if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            newNode.next = current;
+            current.prev.next = newNode;
+            newNode.prev = current.prev;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.isEmpty()) {
+            throw new NoSuchElementException("The list is empty");
+        }
+        for (T l : list) {
+            add(l);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (checkIfIndexIsOutOfBounds(index)) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
```

Comment:
```suggestion
        checkIfIndexIsOutOfBounds(index);
```
move throwing exception to method